File: custom/imageData/processMatrices.m
bbetData = importdata('bb23.txt', ',');
bbhtData = importdata('bb21.txt', ',');
hcetData = importdata('hc23.txt', ',');

bbetIds = bbetData(:, 1);
bbhtIds = bbhtData(:, 1);
hcetIds = hcetData(:, 1);

validIds = intersect(bbetIds, intersect(bbhtIds, hcetIds));
bbetValidIndices = ismember(bbetIds, validIds);
bbhtValidIndices = ismember(bbhtIds, validIds);
hcetValidIndices = ismember(hcetIds, validIds);

bbetData = bbetData(bbetValidIndices, 2:end);
bbhtData = bbhtData(bbhtValidIndices, 2:end);
hcetData = hcetData(hcetValidIndices, 2:end);

numTransforms = length(validIds); 

bbHets = reshape(bbetData', 4, 4, []);
bbHhts= reshape(bbhtData', 4, 4, []);
hcHets = reshape(hcetData', 4, 4, []);


bbHhcs = zeros(4, 4, numTransforms);
for i = 1:numTransforms
    bbHhcs(:, :, i) = bbHets(:, :, i) / hcHets(:, :, i);
end

htHhcs = zeros(4, 4, numTransforms);
for i = 1:numTransforms
    htHhcs(:, :, i) = bbHhts(:, :, i) \ bbHhcs(:, :, i);
end

baselineT = htHhcs(1:3, 4, 1);

avgT = zeros(3,1);
avgR = zeros(3,3);
figure
hold on
skipped = 0;
for i = 1:numCells
    t = htHhcs(1:3, 4, i);
    R = htHhcs(1:3, 1:3, i);
    if norm(t - baselineT) > .03
        skipped = skipped + 1;
        continue
    end
    avgT = avgT + t;
    avgR = avgR + logm(R);
    mArrow3(t, t + R(:,1) * .1, 'color', 'red', 'stemWidth', 0.0002); 
end

avgT = avgT / (numTransforms - skipped)
avgR = expm(avgR / (numTransforms - skipped))
mArrow3(avgT, avgT + avgR(:,1) * .1, 'color', 'blue', 'stemWidth', 0.0002); 

File: custom/imageData/processMatrices.py
import sys
import numpy as np
import scipy as sp

hcValuesSerial = np.loadtxt('leftHandCamera.txt')
bbValuesSerial = np.loadtxt('bumblebee.txt')

if 2 * len(hcValuesSerial) != len(bbValuesSerial):
    sys.exit('Some tags were not detected. Please curate photos.')

numTransformations = len(hcValuesSerial) / 16

hcHets = np.zeros([4, 4, numTransformations])
bbHhts = np.zeros([4, 4, numTransformations])
bbHets = np.zeros([4, 4, numTransformations])

j = 0
for i in xrange(0, len(hcValuesSerial), 16):
    hcHets[:, :, j] = np.reshape(hcValuesSerial[i:i + 16], (4, 4)).T
    j += 1

j = 0
for i in xrange(0, len(bbValuesSerial), 32):
    bbHhts[:, :, j] = np.reshape(bbValuesSerial[i:i + 16], (4, 4)).T
    j += 1

j = 0
for i in xrange(16, len(bbValuesSerial), 32):
    bbHets[:, :, j] = np.reshape(bbValuesSerial[i:i + 16], (4, 4)).T
    j += 1

bbHhcs = np.zeros([4, 4, numTransformations])
for i in xrange(numTransformations):
    bbHhcs[:, :, i] = np.dot(bbHets[:, :, i], np.linalg.inv(hcHets[:, :, i]))

htHhcs = np.zeros([4, 4, numTransformations])
for i in xrange(numTransformations):
    htHhcs[:, :, i] = np.dot(np.linalg.inv(bbHhts[:, :, i]), bbHhcs[:, :, i])

avgT = np.zeros([3, 1])
avgR = np.zeros([3, 3])

baselineT = htHhcs[0:3, 3, 4].reshape(3, 1)
skipped = 0
for i in xrange(numTransformations):
    t = htHhcs[0:3, 3, i].reshape(3, 1)
    if np.linalg.norm(baselineT - t) > .03:
        skipped += 1
        continue
    avgT += htHhcs[0:3, 3, i].reshape(3, 1)
    avgR += htHhcs[0:3, 0:3, i]

avgT = avgT / (numTransformations - skipped)
avgR = avgR / (numTransformations - skipped)

u, s, v = np.linalg.svd(avgR)
avgR = np.dot(u, v)

print avgT
print avgR
